### ISSUE TITLE
Feature request: Add support for High-Throughput JPEG 2000 

### ISSUE THREAD

**Issue by dazzag24** [rordenlab/dcm2niix#897]
> The HTJ2K codec is an order of magnitude faster than JPEG2000, and at least twice as fast as all other DICOM transfer syntaxes. With HealthImaging, applications can leverage HTJ2K with single instruction multiple data (SIMD) instructions, for exceptional image decoding performance. Further, modern browsers can utilize Web Assembly SIMD (WASM-SIMD) to bring industry leading performance to zero-footprint web viewers. 

ref: https://aws.amazon.com/blogs/industries/introducing-aws-healthimaging/

> HTJ2K is the abbreviation for High-Throughput JPEG 2000, which is defined in Part 15 of the JPEG2000 standard (ISO/IEC 15444-15:2019). HTJ2K retains the advanced features of JPEG2000 such as resolution scalability, precincts, tiling, high bit depth, multiple channels, and color space support.
> 
> During import, AWS HealthImaging encodes all [image frames](https://docs.aws.amazon.com/healthimaging/latest/devguide/getting-started-concepts.html#concept-image-frame) in HTJ2K lossless format to deliver consistently fast image display and universal access to HTJ2K’s advanced features. Depending on your programming language, we recommend the following decoding libraries for image frames (pixel data).
> 
>  [OpenJPH](https://github.com/aous72/OpenJPH)
> 
> [OpenJPEG](https://www.openjpeg.org/)
> 
> [openjphpy](https://github.com/UM2ii/openjphpy)
> 
> [pylibjpeg-openjpeg](https://github.com/pydicom/pylibjpeg-openjpeg/)
> 
> [Kakadu Software](https://kakadusoftware.com/)

ref: https://docs.aws.amazon.com/healthimaging/latest/devguide/reference-htj2k.html

Is this something that could make sense for dcm2niix to support?

Thanks

**Comment by neurolabusc** [comment]
@dazzag24 dcm2niix uses OpenJPEG to decode jpeg2000, and OpenJPEG is one of the decoding libraries that supports [HTJ2K](https://docs.aws.amazon.com/healthimaging/latest/devguide/reference-htj2k.html). Therefore, you should have this support if you compile dcm2niix with [OpenJPEG 2.5.0 or later](https://github.com/uclouvain/openjpeg/blob/v2.5.2/NEWS.md#openjpeg-250-may-2022).

**Comment by dazzag24** [comment]
Thanks, I'm currently building with `-DUSE_OPENJPEG=ON ` so perhaps it already works. Will test.

**Comment by dazzag24** [comment]
Does dcm2niix need to be made aware of the new transport syntaxes?  Looks that way.

```
❯  ./dcm2niix -v
Chris Rorden's dcm2niiX version v1.0.20241208  (JP2:OpenJPEG) (JP-LS:CharLS) Clang12.0.0 ARM (64-bit MacOS)
v1.0.20241208

❯ dcm2niix -o ~/Downloads/converted_decomps5 -z y -f %j_%k ~/Desktop/converted_decomps5
Chris Rorden's dcm2niiX version v1.0.20240202  Clang15.0.0 ARM (64-bit MacOS)
Found 473 DICOM file(s)
Warning: Unsupported transfer syntax '1.2.840.10008.1.2.4.201' (see www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage)
Warning: Unsupported transfer syntax '1.2.840.10008.1.2.4.201' (see www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage)
Warning: Unsupported transfer syntax '1.2.840.10008.1.2.4.201' (see www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage)
```

**Comment by neurolabusc** [comment]
I am unable to replicate. Can you share a sample of this transfer syntax with my [institutional email](https://sc.edu/study/colleges_schools/artsandsciences/psychology/our_people/directory/rorden_chris.php), or describe how a DICOM image can be converted to this transfer syntax. It does not seem like [gdcmconv](https://manpages.ubuntu.com/manpages/jammy/man1/gdcmconv.1.html) 3.0.24 that uses OpenJPEG 2.5.3 supports this yet.

**Comment by dazzag24** [comment]
There are a couple of examples here: 
https://github.com/pydicom/pydicom-data/tree/master/data_store/data
see the files beginning "HTJ2K*" which have the following transfer syntaxes

1.2.840.10008.1.2.4.201
1.2.840.10008.1.2.4.203 

See https://github.com/pydicom/pydicom/blob/153a3a9afeb35165ff175ebfbe21eca1a93014fc/src/pydicom/uid.py#L335 for a mapping to the human readable name.

I can send you a full series that uses 1.2.840.10008.1.2.4.201 but its a 140Mb zip so possibly to large to email.

**Comment by dazzag24** [comment]
Hi, Did you receive the example series that I emailed?
Thanks

**Comment by dazzag24** [comment]
I became aware of another set of example image today that are being used by the[ ITK team ](https://github.com/InsightSoftwareConsortium/ITK/pull/5020/commits/1a79eb4134fe3dab4dd339d4ab1067c4569e7944#diff-b7651bcaf4fdd4f408a14e08cf185151b3aa11182de06a961992236c7a27cc6b)

**Comment by neurolabusc** [comment]
@dazzag24 why don't you try out the latest commit to the development branch (v1.0.20241211). You have to compile with the OpenJPEG. The easiest way to do this is with the cmake superbuild:

```bash
git clone --branch development https://github.com/rordenlab/dcm2niix.git
cd dcm2niix
mkdir build && cd build
cmake -DZLIB_IMPLEMENTATION=Cloudflare -DUSE_JPEGLS=ON -DUSE_OPENJPEG=ON ..
make
```
I also added a faster way to build, but you will have to tune the /console/makefile to match the location of your JPEG2000 library:
```bash
git clone --branch development https://github.com/rordenlab/dcm2niix.git
cd dcm2niix\console
make jp2
```

**Comment by dazzag24** [comment]
Hi,

Apologies for the delay. I finally got around to doing some testing on this.  As you know I found an issue with the converted NIfTI files being much smaller but wanted to be sure that it was not a problem with how my DICOM compression code was producing the HTJ2K DICOMs I was using in testing.  I am now confident that this not the case.

Using the test files from Aliza (see above) and from pydicom I also see issues where the output NIfTI file is corrupt.

Using the HTJ2KLossless_08_RGB.dcm file from the [pydicom test data](https://github.com/pydicom/pydicom-data/tree/master/data_store/data)

```
~/Downloads/dcm2niix -o ~/Desktop/niftis -z y -f %j_%k -s y HTJ2KLossless_08_RGB.dcm
Chris Rorden's dcm2niiX version v1.0.20241211  (JP2:OpenJPEG) (JP-LS:CharLS) Clang12.0.0 ARM (64-bit MacOS)
Warning: Instance number (0020,0013) not found: HTJ2KLossless_08_RGB.dcm
Warning: Patient Position (0018,5100) not specified (issue 642).
Warning: Unable to determine manufacturer (0008,0070), so conversion is not tuned for vendor.
Convert 1 DICOM as /Users/xxxxx/Desktop/niftis/_ (640x480x1x1)
Conversion required 0.052217 seconds (0.048659 for core code).
```

The resulting NIfTI looks like this in 3D Slicer:

<img width="983" alt="image" src="https://github.com/user-attachments/assets/f382bb89-05dd-44cd-93dc-2542e3c82188" />

Also with HTJ2K-YBR_ICT.dcm file from the [pydicom test data](https://github.com/pydicom/pydicom-data/tree/master/data_store/data)

```
~/Downloads/dcm2niix -o ~/Desktop/niftis -z y -f %j_%k -s y Aliza-DICOM-viewer-HTJ2k/HTJ2K-YBR_ICT.dcm
Chris Rorden's dcm2niiX version v1.0.20241211  (JP2:OpenJPEG) (JP-LS:CharLS) Clang12.0.0 ARM (64-bit MacOS)
Warning: Patient Position (0018,5100) not specified (issue 642).
Warning: Unable to determine manufacturer (0008,0070), so conversion is not tuned for vendor.
Convert 1 DICOM as /Users/xxxxx/Desktop/niftis/1.2.826.0.1.3680043.9.4245.6602037944901915650477181237096105672_1.2.826.0.1.3680043.9.4245.7964578538842334580361701724537168429 (1920x1080x1x1)
Conversion required 0.271147 seconds (0.249147 for core code).
```

The resulting NIfTI looks like this in 3D Slicer:

<img width="989" alt="image" src="https://github.com/user-attachments/assets/d86a1569-a496-4d09-877d-dd1109543992" />

NIfTIs from my own test dataset I see output like this:

<img width="980" alt="image" src="https://github.com/user-attachments/assets/8e508710-ed1e-4e7e-bce5-e3c4a0c727f7" />

I'm fairly confident that my HTJ2K DICOMs are valid as I have code that compares the pixel array from the original DICOMs with the decompressed pixel arrays and they seem valid (e,g I have test code that takes an original DICOM, compresses it and then opens the compressed file using pydicom accesses the pixel_array object).  

Furthermore my HTJ2K compressed DICOMs are correctly rendered in both 
https://unpkg.com/dcmjs-imaging@0.1.26/build/index.html 
and
https://chafey.github.io/openjphjs/test/browser/index.html

Please let me know if you want anymore info.

Thanks

**Comment by neurolabusc** [comment]
@dazzag24 you are correct, the current dcm2niix cmake uses [OpenJPEG 2.1](https://github.com/rordenlab/dcm2niix/blob/master/SuperBuild/External-OPENJPEG.cmake) while high throughput (HTJ2K) decoding requires  [OpenJPEG 2.5.0 or later](https://github.com/uclouvain/openjpeg/blob/v2.5.3/NEWS.md#openjpeg-250-may-2022). You can use the `make jp2` from the console folder to see that dcm2niix works correctly with these images, but this requires you to manually install a recent OpenJPEG version. 

@ningfei would it be possible to update your cmake library for the latest stable [OpenJPEG](https://github.com/uclouvain/openjpeg/blob/v2.5.3/NEWS.md#openjpeg-253-dec-2024)?

**Comment by dazzag24** [comment]
Using an exe compiled using `make jp2` yields sensible looking results upon initial inspection:

./dcm2niix -o ~/Desktop/niftis -z y -f %j_%k -s y Aliza-DICOM-viewer-HTJ2k/HTJ2K-YBR_ICT.dcm
Chris Rorden's dcm2niiX version v1.0.20241211  (JP2:OpenJPEG) Clang16.0.0 ARM (64-bit MacOS)
Warning: Patient Position (0018,5100) not specified (issue 642).
Warning: Unable to determine manufacturer (0008,0070), so conversion is not tuned for vendor.
Convert 1 DICOM as /Users/xxxxx/Desktop/niftis/1.2.826.0.1.3680043.9.4245.6602037944901915650477181237096105672_1.2.826.0.1.3680043.9.4245.7964578538842334580361701724537168429 (1920x1080x1x1)
Conversion required 0.293713 seconds (0.285138 for core code).

<img width="880" alt="image" src="https://github.com/user-attachments/assets/00ef1887-5c94-463e-a636-1230e9e6201b" />

**Comment by ningfei** [comment]
will have a look when I find the time.

**Comment by dazzag24** [comment]
For now I have added some extra steps to my Dockerfile that downloads and compiles 
https://github.com/uclouvain/openjpeg/archive/refs/tags/v2.5.3.tar.gz 
locally.  I've had to add a slightly hacky sed replace command to replace the hardcoded `/opt/homebrew/Cellar/openjpeg/2.5.3` in the console makefile, but it works.

Thanks

**Comment by dazzag24** [comment]
Quick question.  I normally compile with
`
cmake -DBATCH_VERSION=ON -DZLIB_IMPLEMENTATION=Cloudflare -DUSE_JPEGLS=ON `

By using `make jp2` in the console directory am I missing out on the zlib optimisation and JPEGLS support?

**Comment by neurolabusc** [comment]
I have added a commit to the development branch that provides arguments to `make` can combine JPEG2000 and JPEG-LS support. DICOM supports [different types of JPEG](https://crnl.readthedocs.io/jpeg_formats/index.html). dcm2niix always includes inbuilt support for lossless JPEG, requires CharLS for JPEG-lossless, and requires OpenJPEG for JPEG2000.

| command           | features                 |
| ----------------- | ------------------------ |
| make              | miniz compressor         |
| make jp2          | JP2000 support           |
| JPEGLS=1 make     | JP-LS support            |
| JPEGLS=1 make jp2 | JP2000 and JP-LS support |

With regards to creating compressed NIfTIs, dcm2niix can be compiled to use one of 3 internal compressors: [miniz](https://github.com/richgel999/miniz) (slightly slower than zlib), system zlib, or [cloudflare zlib](https://github.com/cloudflare/zlib) (about twice as quick as zlib). The `make` command only compiles with the native compressor, the `cmake` provides arguments to make any compile. Regardless of the internal compressor, you will always get optimal compression speed by installed `pigz` and using it as your external compressor. Assuming a modern version of pigz, the fastest way to create gzipped files is the `optimal` (`-z o`) argument for dcm2niix, `dcm2niix -z o /path/to/DICOMs`. This method transfers uncompressed memory directly to pigz, rather than creating an uncompressed file and having pigz compress that file. The major strength and limitation of pigz is that it is multi-threaded, so it may not be suitable to instances where you want all processing constrained to a single thread. Another weakness of pigz is that assumes Unix style UTF8 file names, [Windows UTF16 filenames can cause issues](https://github.com/madler/pigz/pull/77) - e.g. if you have a folder name with non-Latin characters.

**Comment by neurolabusc** [comment]
The latest commit introduces the option to use the system zlib during compilation with `make`. This approach results in a slightly smaller executable because the compression library is not statically included. However, since it dynamically links to zlib, the zlib library must be installed on your system for the executable to function properly.

To compile with the system zlib, use the following command:
```bash
JPEGLS=1 ZLIB=1 make jp2
```

**Comment by dazzag24** [comment]
Thanks, I was already using `-z y` which i think is equivalent to `-z o`. I've added the `JPEGLS=1` line.

**Comment by neurolabusc** [comment]
@dazzag24 note that while both the arguments `-z y` and `-z o` can use pigz to accelerate compression, they are slightly different. The `-z y` will write each NIfTI file to disk uncompressed and then compress the file, analogous to running the command `pigz --no-time -n -f -6 /path/to/image.nii'`. On the other hand, using `-z o` on a Unix computer will pipe the raw data directly to pigz, without creating the intermediate uncompressed file. This piped operation requires you have pigz 2.3.4 (released Oct 2, 2016) or later. The difference is negligible for fast local SSD storage, but can be substantial if you are saving NIfTI images to networked storage.

So the options are:

| argument | notes |
| --------- | -----------|
| -z n          | create uncompressed NIfTIs |
| -z y          | use pigz if found (intermediate file),  else use internal compressor |
| -z o         | use pigz if found (piped data),  else use internal compressor |
| -z i          | use internal compressor |

If you run dcm2niix without any arguments, the help file will tell you if it was compiled with miniz or zlib as an internal compressor. HOwever, it does not distringuish between the system zlib and cloudflare zlib as they are both zlib compatible libraries.
```
 -z : gz compress images (y/o/i/n/3, default n) [y=pigz, o=optimal pigz, i=internal:zlib, n=no, 3=no,3D]
 -z : gz compress images (y/o/i/n/3, default n) [y=pigz, o=optimal pigz, i=internal:miniz, n=no, 3=no,3D]
```

**Comment by dazzag24** [comment]
Thanks for the detailed explanation. Would be great to add this to some
docs. I don't think the current README.md goes into that level of detail?

**Comment by neurolabusc** [comment]
[I have updated the wiki](https://www.nitrc.org/plugins/mwiki/index.php?title=dcm2nii:MainPage#General_Usage)

**Comment by soham1300** [comment]
Any updates regarding this issue? I am also encountering the same error: 'Warning: Unsupported transfer syntax '1.2.840.10008.1.2.4.91'.' I am using the @niivue/dcm2niix library.

**Comment by neurolabusc** [comment]
@soham1300 dcm2niix is modular, so the compression types (transfer syntaxes) depend on the libraries it is compiled with. As described in the manual, [dcm2niix does support 1.2.840.10008.1.2.4.91](https://www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage#DICOM_Transfer_Syntaxes_and_Compressed_Images) if compiled with the OpenJPEG library. 

For natively compiled code, this is included with the binary releases, and you can make your own command line version using the [recommended compilation:](https://github.com/rordenlab/dcm2niix)

```bash
git clone https://github.com/rordenlab/dcm2niix.git
cd dcm2niix
mkdir build && cd build
cmake -DZLIB_IMPLEMENTATION=Cloudflare -DUSE_JPEGLS=ON -DUSE_OPENJPEG=ON ..
make
```

However,  @niivue/dcm2niix is a [WebAssembly release for use with JavaScript](https://github.com/rordenlab/dcm2niix/tree/master/js), and I am not sure what libraries it includes. I do think porting libraries to JavaScript may be challenging. Perhaps @hanayik can provide details.

As an aside, [1.2.840.10008.1.2.4.91](https://www.dicomlibrary.com/dicom/transfer-syntax/) allows lossy compression. Unlike old JPEG, where the lossyness leads to visible blocky artifacts, the JPEG2000 lossy is very subtle and looks like a loss of high frequency blurriness. My own opinion is that blurriness can also be associated with some pathology (e.g. medial temporal lobe epilepsy). I would personally avoid lossy compression for medical data.

**Comment by dazzag24** [comment]
Hi,

On this subject of lossy compression causing blurriness. I'd be interested
to know if you think this also applies to the newer HTJ2K transfer syntaxes
such as 1.2.840.10008.1.2.4.203 when using a conservative quantization step
value of say 0.00039?

You can test this out [here](
https://chafey.github.io/openjphjs/test/browser/index.html) and can use the
"visualise deltas" checkbox to see the difference.  To my radiologically
untrained eye you can achieve impressive compression results with very
minimal change to the image. For our own use case we ALWAYS preserve the
original data provided BUT there is a use case where being able to serve up
significantly smaller DICOMs to a user who is viewing many scans (for say a
quick QC workflow) can make this a much better user experience.

**Comment by neurolabusc** [comment]
@dazzag24 your question is out of scope for the dcm2niix issues forum. I have provided my own personal opinion that I am wary of lossy compression for medical compression. I agree that any tool that interpolates images to a different scaling will also cause artifacts. The J2K compression does achieve great compression with very little loss in fidelity that can be distinguished by the human eye. Typical scans have relatively low SNR, and there reconstruction also does mean each sample is not discrete. I am really not an expert in medical imaging compression. I tend to avoid lossy compression and interpolation at reconstruction, but these are based on my personal biases not expertise. Your choice depends on your application.

**Comment by dazzag24** [comment]
Thanks.  Understood.

**Comment by hanayik** [comment]
> However, @niivue/dcm2niix is a [WebAssembly release for use with JavaScript](https://github.com/rordenlab/dcm2niix/tree/master/js), and I am not sure what libraries it includes. I do think porting libraries to JavaScript may be challenging. Perhaps @hanayik can provide details.
> 

@soham1300 and @neurolabusc , The WASM build of dcm2niix was compiled with [this line](https://github.com/rordenlab/dcm2niix/blob/98ecd0bec9a1dfabd77b2be8c54c3bae97b92a55/console/makefile#L68). The `JPEGLS` environment variable was not set, so JPEGLS was not added. In fact, those flags are not even part of the current WASM target in the make file, so the environment variables won't do anything. This will need to be modified. 

@neurolabusc , This could be a potential use of multiple `exports` fields in our package.json file, so that downstream developers choose to use dcm2niix WASM builds with or without JPEGLS support. We could just compile both versions and include them in the package. 

Giving developers the ability to do:

```js
import { Dcm2niix } from '@niivue/dcm2niix/with-jpegls'
```

or 

```js
import { Dcm2niix } from '@niivue/dcm2niix' // default, without jpegls
```

**Comment by soham1300** [comment]
Thank you, @neurolabusc and @hanayik, for the detailed explanations and insights!  

From the discussion, it appears that the current WASM build of `@niivue/dcm2niix` does not support the `1.2.840.10008.1.2.4.91` transfer syntax due to the absence of OpenJPEG or JPEGLS integration in the build process. I now understand that addressing this limitation would involve either modifying the WASM build process or waiting for a future update that includes multiple builds with export fields.  

To clarify my next steps:  
1. Would it be feasible for me to fork and modify the current WASM build to include OpenJPEG or JPEGLS? If so, I would greatly appreciate any guidance on how to approach this.  
2. Are there any plans to introduce multiple builds of `@niivue/dcm2niix` (e.g., with and without JPEGLS support) in the near future?

**Comment by hanayik** [comment]
@neurolabusc and @soham1300 , I have created a new PR (https://github.com/rordenlab/dcm2niix/pull/906) that adds a jpegls build for the WASM target. Once the PR is approved and merged, I will publish a new NPM package. 

The new package will include both WASM builds (with, and without jpegls support) so that developers can choose the dcm2niix wasm build that they want in their application. 

The benefit of this approach is that both WASM builds are not going to bloat the apps they are bundled into. Only the relevant files for the imported build will get included by JavaScript module bundlers.

**Comment by hanayik** [comment]
@neurolabusc

I have created a new PR that adds jpegls and openjpeg support to the WASM build of dcm2niix. This will be useful for developers who wish to use dcm2niix in browser environments will full available jpeg support. 

More details are in the PR: https://github.com/rordenlab/dcm2niix/pull/910

**Comment by ningfei** [comment]
I just updated OpenJPEG to v2.5.3 in the CMake external project. The CI build will have the updated version from now on.

**Comment by neurolabusc** [comment]
@ningfei thanks! @dazzag24 closing this issue. The new capabilities are part of the current development release and will be part of the next stable release (Spring 2025). In the meantime, you can compile the development release:

```bash
git clone --branch development https://github.com/rordenlab/dcm2niix.git
cd dcm2niix
mkdir build && cd build
cmake -DZLIB_IMPLEMENTATION=Cloudflare -DUSE_JPEGLS=ON -DUSE_OPENJPEG=ON ..
make
```
